[Clang] [C++26] Expansion Statements (P1306R5)

Sirraide opened the pull request:
This is the first part of the expansion statements ([P1306R5](https://wg21.link/P1306R5)) implementation. First things first, once everything has been approved, I’m going to merge the patches top-down into a single huge patch, and then merge that patch so we only end up with a single commit on trunk. Also, for the sake of avoiding too much rebasing, I’m only going to run `clang-format` after everything has been merged into a single patch.

This patch it introduces all of the AST nodes that are needed by various parts of the implementation. I know that this is still quite a bit of code... but it’s mostly just AST node boilerplate and splitting it up further wouldn’t have helped much I think.

In addition to the AST nodes, all of the decl-context specific code such as the many places where we need to ignore surrounding expansion statement contexts are also added here so we have all of that in one place.

Current state of CWG issues around expansion statements (updated regularly):
- [x] ~~[CWG 3043](https://wg21.link/cwg3043): I still need to implement this one.~~ Will be a follow-up patch; see https://github.com/llvm/llvm-project/pull/169680#issuecomment-4093665027
- [x] [CWG 3044](https://wg21.link/cwg3044): Implemented
- [x] [CWG 3045](https://wg21.link/cwg3045): Implemented
- [x] [CWG 3048](https://wg21.link/cwg3048): Implemented (sort of; we still create a DecompositionDecl if there are 0 elements, but that doesn't seem to cause any issues)
- [x] [CWG 3061](https://wg21.link/cwg3061): Implemented
- [x] [CWG 3123](https://wg21.link/cwg3123): Implemented; we currently perform overload resolution for begin and end and only if both are found do we consider it to be iterating.
- [x] [CWG 3131](https://wg21.link/cwg3131): Implemented
- [x] [CWG 3140](https://wg21.link/cwg3140): Implemented
- [x] [CWG 3149](https://wg21.link/cwg3149): Implemented

There is some follow up work to be done after this lands, which is tracked in #207776.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 1)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 1)

Sirraide commented:
Sorry for wreaking havoc on everyone’s inbox w/ these, but hopefully reviewing this is somewhat more manageable now...

Sirraide commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,120 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// the declaration of the variable 'x' as well as the expansion-initializer
+/// 'a'.
+///
+/// After expansion, we end up with a 'CXXExpansionStmtInstantiation' that
```
Ok, so that is explained in more detail in the `CXXExpansionStmtInstantiation` documentation I believe, but the *outermost* compound statement explicitly _does not exist_ in the AST (only the inner ones shown here do); it pretty much is already as you’re describing (i.e. we handle the outer scope in codegen).

shafik replied:
I really like having the comment here but I wonder if there is a way we can automate pulling sections into the internals document from comments here and other places.

erichkeane commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,120 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// well as a 'CXXExpansionStmtInstantiation'. These two members correspond to
+/// distinct representations of the expansion statement: the former is used
+/// prior to expansion and contains all the parts needed to perform expansion;
+/// the latter holds the expanded/desugared AST nodes that result from the
```
sensible, i think that is reasonable.  Can we put that in the comment?

erichkeane commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -5499,6 +5499,165 @@ class BuiltinBitCastExpr final
   }
 };
 
+/// Represents an expansion-init-list of an enumerating expansion statement.
```
Ooof, thats an awful name.  Yeah, a quick example would be lovely.

erichkeane commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,120 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// }
+/// \endverbatim
+///
+/// The outer braces shown above are implicit; we don't actually create another
```
Yeah, though this says we don't create a compound stmt, though I'm sure I saw one?

Sirraide commented:
Oh yeah, and `CXXExpansionInitListExpr` is also gone now and everything just uses `InitListExpr`.

shafik reviewed (commented):
First things first, your summary should list the paper number and ideally have a link to it as well Alternatively, it could go in the title.

The main point is that the Title/Summary should provide all the information needed for a reviewer to review the PR w/o having to go searching for more information. Secondly, we want the PR to be accessible for as many reviewers as possible and providing this information allows non-WG21 folks to effectively review this PR. We need to do a better job here.

shafik commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,127 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// need them to be declaration *contexts*, because the DeclContext is used to
+/// compute the 'template depth' of entities enclosed therein. In particular,
+/// the 'template depth' is used to find instantiations of parameter variables,
+/// and a lambda enclosed within an expansion statement cannot compute its
```
```suggestion
/// the 'template depth' is used to find instantiations of parameter variables.
/// A lambda enclosed within an expansion statement cannot compute its
```

shafik commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,127 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// \see CXXExpansionStmtInstantiation
+class CXXExpansionStmtDecl : public Decl, public DeclContext {
+  CXXExpansionStmtPattern *Expansion = nullptr;
+  NonTypeTemplateParmDecl *IndexNTTP = nullptr;
```
Why do we need this? I am having a hard time seeing the larger context of this.

Sirraide replied:
The comment goes into more detail about this, but tl;dr this is needed because we need expansion statements to introduce a new `DeclContext` (e.g. so we know that we’re in a dependent context), and every `DeclContext` must also be a `Decl`. I believe `RequiresExprBodyDecl` exists for similar reasons.

cor3ntin replied:
Maybe the suggestion is to document `IndexNTTP` - I certainly would agree with that

Sirraide replied:
There’s some documentation in that comment:
```
/// Additionally, there is a 'NonTypeTemplateParmDecl', which is a template
/// parameter that serves as the expansion index, e.g. during the N-th
/// expansion, it is set to 'N'. See the documentation of
/// 'CXXExpansionStmtPattern', for more information on how this is used.
```
and some more in the one on `CXXExpansionStmtPattern`; do you want me to add more?

erichkeane reviewed (approved):
I think I'm ok with this one.  There's a lot of stuff in this patch, but I did as good of a review as I could, and it seems reasonable.

Sirraide commented:
ping

(Erich has already reviewed the patches, and Eli has looked at the codegen changes, but we’d still like some more people to take a look at it before merging)

shafik commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,127 @@ class TemplateParamObjectDecl : public ValueDecl,
+///
+/// Apart from a template parameter list that contains a template parameter used
+/// as the expansion index, this node contains a 'CXXExpansionStmtPattern' as
+/// well as a 'CXXExpansionStmtInstantiation'. These two members correspond to
```
Can you also expand upon `NonTypeTemplateParmDecl`.

Sirraide replied:
Done

shafik commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -5554,6 +5554,52 @@ class CXXReflectExpr : public Expr {
+class CXXExpansionSelectExpr : public Expr {
+  friend class ASTStmtReader;
+
+  enum SubExpr { RANGE, INDEX, COUNT };
```
INDEX here feels a little abstract but maybe later comments will help.

Ok looks like this is covered in `CXXExpansionStmtPattern` comment, maybe a forward reference would be helpful, "if you want to know more see ...."

Sirraide replied:
Added a reference to that

shafik commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,520 @@ class CoreturnStmt : public Stmt {
+/// [stmt.ranged]).
+///
+/// The expression used to compute the size of the expansion is not stored and
+/// is only created at the moment of expansion.
```
This feels rather sparse compared to the previous entry.

Sirraide replied:
Added some more documentation, but I also didn’t just want to copy-paste the entirety of [stmt.expand] here...

shafik commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,520 @@ class CoreturnStmt : public Stmt {
+///
+/// This will be instantiated as either an iterating or destructuring expansion
+/// statement. Dependent expansion statements can never be enumerating, even if
+/// the expansion size is dependent because the expression-list contains a pack.
```
The details in the first case are really great and the last three are very sparse.

shafik commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,520 @@ class CoreturnStmt : public Stmt {
+    COUNT_Enumerating = FIRST_CHILD_STMT,
+
+    // Dependent expansion initializer.
+    EXPANSION_INITIALIZER = FIRST_CHILD_STMT,
```
The use of `FIRST_CHILD_STMT` over again here deserves a detailed comment.

Sirraide replied:
Added a comment; hopefully it’s clearer now

shafik commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -1347,6 +1348,11 @@ void DeclPrinter::VisitClassTemplatePartialSpecializationDecl(
 }
 
+void DeclPrinter::VisitCXXExpansionStmtDecl(const CXXExpansionStmtDecl *D) {
+  D->getExpansionPattern()->printPretty(Out, nullptr, Policy, Indentation, "\n",
```
```suggestion
  D->getExpansionPattern()->printPretty(Out, /*PrinterHelper*=/nullptr, Policy, Indentation, "\n",
```

shafik commented on clang/lib/AST/DeclTemplate.cpp:
```diff
@@ -1784,3 +1785,22 @@ const Decl &clang::adjustDeclToTemplate(const Decl &D) {
+}
+CXXExpansionStmtDecl *
+CXXExpansionStmtDecl::CreateDeserialized(ASTContext &C, GlobalDeclID ID) {
+  return new (C, ID) CXXExpansionStmtDecl(nullptr, SourceLocation(), nullptr);
```
```suggestion
  return new (C, ID) CXXExpansionStmtDecl(/*DC=*/nullptr, SourceLocation(), /*NTTP*=/nullptr);
```

shafik commented on clang/lib/AST/TextNodeDumper.cpp:
```diff
@@ -1512,6 +1516,30 @@ void clang::TextNodeDumper::VisitCoreturnStmt(const CoreturnStmt *Node) {
+  case CXXExpansionStmtPattern::ExpansionStmtKind::Dependent:
+    OS << " dependent";
+    break;
+  }
```
llvm_unreachable?

Sirraide replied:
I don’t think we usually do that in the TextNodeDumper but I can return in the switch statement and add unreachable at the end sure

shafik commented on clang/lib/Sema/IdentifierResolver.cpp:
```diff
@@ -96,9 +96,10 @@ IdentifierResolver::~IdentifierResolver() {
-/// isDeclInScope - If 'Ctx' is a function/method, isDeclInScope returns true
-/// if 'D' is in Scope 'S', otherwise 'S' is ignored and isDeclInScope returns
-/// true if 'D' belongs to the given declaration context.
+/// isDeclInScope - If 'Ctx' is a function/method/expansion statement,
```
This is kind of hard to read, I don't have a suggestion other than not this.

Sirraide replied:
Yeah, combining slashes and phrases that contain more than one word is usually a bit awful; I'll see if I can improve this

shafik commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -7645,7 +7645,6 @@ static void CheckIfAnyEnclosingLambdasMustCaptureAnyPotentialCaptures(
     Expr *const FE, LambdaScopeInfo *const CurrentLSI, Sema &S) {
 
   assert(!S.isUnevaluatedContext());
-  assert(S.CurContext->isDependentContext());
```
Why this change?

Sirraide replied:
Generally, I had to remove a number of assertions in the template instantiator and around dependent contexts; I don’t remember why this one was removed in particular, but the general answer is that it’s now possible to perform template instantiation inside a function w/o instantiating the entire function.

I can re-add this and see what breaks if you want to know why in particular this assertion needs to be removed

shafik commented on clang/lib/Sema/SemaLambda.cpp:
```diff
@@ -100,8 +100,9 @@ static inline UnsignedOrNone getStackIndexOfNearestEnclosingCaptureReadyLambda(
-    if (IsCapturingVariable &&
-        VarToCapture->getDeclContext()->Equals(EnclosingDC))
+    if (IsCapturingVariable && VarToCapture->getDeclContext()
+                                   ->getEnclosingNonExpansionStatementContext()
```
I see you have throw these (`getEnclosingNonExpansionStatementContext`) in a bunch of places and it is not clear to me how we can ensure we did not miss spots that could blow up later on.

Sirraide replied:
> how we can ensure we did not miss spots that could blow up later om

I’ve thought about this too, and I don’t think we can candidly.

cor3ntin replied:
In a few places we skip over requires clause for similar reason.
Can you have an expansion in a require clause, such that we need to skip both?

Sirraide replied:
I don’t think so; requires clauses can’t contain arbitrary statements afaik (unless it’s in a lambda, but in that case the enclosing decl context would be the lambda itself, which is probably what we want)

shafik commented on clang/lib/Sema/SemaLambda.cpp:
```diff
@@ -190,11 +192,6 @@ UnsignedOrNone clang::getStackIndexOfNearestEnclosingCaptureCapableLambda(
     return NoLambdaIsCaptureCapable;
 
   const unsigned IndexOfCaptureReadyLambda = *OptionalStackIndex;
-  assert(((IndexOfCaptureReadyLambda != (FunctionScopes.size() - 1)) ||
```
Why this change?

Sirraide replied:
See [this comment](https://github.com/llvm/llvm-project/pull/169680/changes/BASE..8e338ce8eb8458ed20dc4a006ee15278aac96dd1#r3096865117), but  essentially, we can now instantiate code inside a lambda even if the lambda itself is not a template / not generic

shafik reviewed (commented):
I think I did a pretty complete run through but this is quite large, it is a bit daunting considering how many more there are.

Sirraide commented:
ping

cor3ntin commented on clang/include/clang/AST/DeclTemplate.h:
```diff
@@ -3343,6 +3343,131 @@ class TemplateParamObjectDecl : public ValueDecl,
+/// \see CXXExpansionStmtPattern
+/// \see CXXExpansionStmtInstantiation
+class CXXExpansionStmtDecl : public Decl, public DeclContext {
+  CXXExpansionStmtPattern *Expansion = nullptr;
```
Maybe it would be clearer to call the variable `Pattern` ?

Sirraide replied:
Done; I’ll wait a bit before I push the changes because github sends out 10 emails every time I do...

cor3ntin commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,581 @@ class CoreturnStmt : public Stmt {
+  };
+
+private:
+  ExpansionStmtKind PatternKind;
```
We could put that in a `StmtBitfields` - i don't know if expansion statement would be common enough to bother though

Sirraide replied:
I don’t think it would make a difference in this case: we have 4 bytes available here because of the 3 `SourceLocation`s, but if we end up removing one then I guess we could move this into `StmtBitfields`

cor3ntin commented on clang/include/clang/AST/StmtCXX.h:
```diff
@@ -524,6 +525,581 @@ class CoreturnStmt : public Stmt {
+  ExpansionStmtKind PatternKind;
+  SourceLocation LParenLoc;
+  SourceLocation ColonLoc;
+  SourceLocation RParenLoc;
```
Why do we keep track of the left paren and not the keyword?

Sirraide replied:
Iirc Erich said that we shouldn’t keep track of the keyword; afaik we usually try to have source locations for everything in the AST, but if we don’t care about that then I can remove it if it’s not used

cor3ntin replied:
@erichkeane ???

cor3ntin replied:
Ping on that

cor3ntin commented on clang/include/clang/Sema/Scope.h:
```diff
@@ -196,6 +196,16 @@ class Scope {
   /// declared in this scope.
   unsigned short PrototypeIndex;
 
+  /// IsExpansionStmtScope - This is the scope corresponding to a C++26
+  /// expansion statement.
+  ///
+  /// FIXME: This should be part of ScopeFlags, but we're out of bits, so we
+  /// need to update every place that uses 'unsigned' to hold scope flags. We
+  /// should probably redefine ScopeFlags as an 'enum class : uint64_t' and
+  /// use LLVM_MARK_AS_BITMASK_ENUM() and friends so we can continue to '|'
+  /// scope flags together.
+  bool IsExpansionStmtScope;
+
```
Can we do that _before_ this PR?

Sirraide replied:
I mean, I can look into that I suppose, but not sure how much work that would end up being

Sirraide replied:
#198436 should free up a bit that we can use for this

Sirraide replied:
I candidly don’t want to rebase all of the patches individually, so... do you mind if we switch to using this bit in a follow-up patch?

Sirraide replied:
Added #207774 to track this so we don’t forget about it.

cor3ntin commented on clang/lib/Sema/SemaStmtAttr.cpp:
```diff
@@ -38,6 +38,18 @@ static Attr *handleFallThroughAttr(Sema &S, Stmt *St, const ParsedAttr &A,
     return nullptr;
   }
 
+  // CWG 3045: The innermost enclosing switch statement of a fallthrough
+  // statement S shall be contained in the innermost enclosing expansion
+  // statement (8.7 [stmt.expand]) of S, if any.
```
Will this be covered by a DR test?

Sirraide replied:
I wasn’t planning to add one but I do have a test for it in patch 9 I think

cor3ntin replied:
ideally we should have a dr test for everything so that reflects the completeness of your work
https://clang.llvm.org/cxx_dr_status.html

Sirraide replied:
Makes sense; do you mind if we defer that to a follow-up patch?

Sirraide replied:
Added #207790 to track this so we don’t forget about it.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 2: Parsing and Parser Tests) (#169681)

This patch implements parsing of expansion statements and expansion
initialiser lists. I’ve also had to implement sema for
`CXXExpansionStmtDecl` because parsing doesn’t really work without that
node.

Sirraide committed:
Merge remote-tracking branch 'origin/main' into users/Sirraide/expansion-stmts-1-ast

Sirraide committed:
Fix patch after changes in main

Sirraide committed:
clang-format

Sirraide committed:
Move implementation into .cpp file
